InaccessibleObjectException in JDK 11

Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make public long com.sun.management.internal.OperatingSystemImpl.getMaxFileDescriptorCount() accessible: module jdk.management does not "opens com.sun.management.internal" to unnamed module @22eeefeb

--- a/src/main/java/ai/apptuit/metrics/jinsight/modules/jvm/FileDescriptorMetrics.java
+++ b/src/main/java/ai/apptuit/metrics/jinsight/modules/jvm/FileDescriptorMetrics.java
@@ -37,15 +37,17 @@ class FileDescriptorMetrics implements MetricSet {
 
   private static final Logger LOGGER = Logger.getLogger(FileDescriptorMetrics.class.getName());
 
-  private OperatingSystemMXBean osMxBean;
+  private final Class unixOSMBeanClass;
+  private final OperatingSystemMXBean osMxBean;
 
-  public FileDescriptorMetrics() {
+  public FileDescriptorMetrics() throws ClassNotFoundException {
     this(getOperatingSystemMXBean());
   }
 
-  public FileDescriptorMetrics(OperatingSystemMXBean osMxBean) {
-
+  public FileDescriptorMetrics(OperatingSystemMXBean osMxBean) throws ClassNotFoundException {
     this.osMxBean = osMxBean;
+    unixOSMBeanClass = Class.forName("com.sun.management.UnixOperatingSystemMXBean");
+    unixOSMBeanClass.cast(this.osMxBean);
   }
 
   public Map<String, Metric> getMetrics() {
@@ -59,7 +61,8 @@ class FileDescriptorMetrics implements MetricSet {
 
   private Long getMetricLong(String name) {
     try {
-      return invoke(name);
+      final Method method = unixOSMBeanClass.getDeclaredMethod(name);
+      return (Long) method.invoke(osMxBean);
     } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
       LOGGER.log(Level.SEVERE,
           "Error fetching file descriptor metrics from OperatingSystemMXBean", e);
@@ -67,10 +70,4 @@ class FileDescriptorMetrics implements MetricSet {
     }
   }
 
-  private long invoke(String name)
-      throws NoSuchMethodException, IllegalAccessException, InvocationTargetException {
-    final Method method = osMxBean.getClass().getDeclaredMethod(name);
-    method.setAccessible(true);
-    return (Long) method.invoke(osMxBean);
-  }
 }

--- a/src/main/java/ai/apptuit/metrics/jinsight/modules/jvm/JvmMetricSet.java
+++ b/src/main/java/ai/apptuit/metrics/jinsight/modules/jvm/JvmMetricSet.java
@@ -53,7 +53,12 @@ public class JvmMetricSet implements MetricSet {
   public JvmMetricSet() {
     registerSet("jvm.buffer_pool", new BufferPoolMetrics());
     registerSet("jvm.classes", new ClassLoadingGaugeSet());
-    registerSet("jvm.fd", new FileDescriptorMetrics());
+    try {
+      registerSet("jvm.fd", new FileDescriptorMetrics());
+    } catch (ClassNotFoundException|RuntimeException e) {
+      LOGGER.log(Level.SEVERE, "Could not load FileDescriptorMetrics."
+              + " File descriptor metrics will not be available.", e);
+    }
     registerSet("jvm.gc", new GarbageCollectorMetrics());
     registerSet("jvm.memory", new MemoryUsageMetrics());
     registerSet("jvm.threads", new ThreadStateMetrics());

